examples/streaming/streamingkmeans/src/main/scala/io/gearpump/streaming/examples/streamingkmeans/ClusterDistribution.scala

```diff
@@ -0,0 +1,143 @@
+  private[streamingkmeans] var currentNumber = 0
+
+
+  /**
+   * init center randomly
+   */
+  private[streamingkmeans] def initCenter(): Unit = {
+    val random = new Random()
+    for (i <- center.indices) {
+      center.update(i, random.nextGaussian())
+    }
+  }
+
+  /**
+   * The update algorithm uses the "mini-batch" KMeans rule,
+   * generalized to incorporate forgetfullness (i.e. decay).
+   * The update rule (for each cluster) is:
+   *
+   * {{{
+   * c_t+1 = [(c_t * n_t * a) + (x_t * m_t)] / [n_t + m_t]
+   * n_t+t = n_t * a + m_t
+   * }}}
+   *
+   * Where c_t is the previously estimated centroid for that cluster,
+   * n_t is the number of points assigned to it thus far, x_t is the centroid
+   * estimated on the current batch, and m_t is the number of points assigned
+   * to that centroid in the current batch.
+   *
+   * The decay factor 'a' scales the contribution of the clusters as estimated thus far,
+   * by applying a as a discount weighting on the current point when evaluating
+   * new incoming data. If a=1, all batches are weighted equally. If a=0, new centroids
+   * are determined entirely by recent data. Lower values correspond to
+   * more forgetting.
+   */
+  private[streamingkmeans] def updateCenter(): Unit = {
+    if (0 == currentNumber) {
+      return
+    }
+
+    val newCenter: Array[Double] = new Array[Double](dimension)
+    for (i <- newCenter.indices) {
+      var sum = 0.0
+      for (point <- points) {
+        sum += point(i)
+      }
+      sum /= currentNumber
+      newCenter.update(i, sum)
+    }
+
+    for (i <- center.indices) {
+      center.update(i,
+        (center(i) * previousNumber * decayFactor + newCenter(i) * currentNumber)
+          / (previousNumber + currentNumber))
+    }
+  }
+
+  private[streamingkmeans] def getDistance(point: List[Double]): Double = {
+    var distance = 0.0
+    for (i <- 0 until dimension) {
+      distance += ((point(i) - center(i)) * (point(i) - center(i)))
+    }
+    Math.sqrt(distance)
+  }
+
+  override def onStart(startTime: StartTime): Unit = {
+    initCenter()
+  }
+
+  override def onNext(msg: Message): Unit = {
+    if (null == msg) {
+      return
+    }
+
+    val message = msg.msg.asInstanceOf[ClusterMessage]
+
+    message match {
+      case InputMessage(point) =>
+        if (isBegin) {
+          isBegin = false
+          output(new Message((taskContext.taskId.index, getDistance(point), point)))
+        } else {
+          dataQueue.put(point)
+        }
+      case ResultMessage(taskId, point, doCluster) =>
+        if (taskContext.taskId.index == taskId) {
+          points += point
+          currentNumber += 1
+        }
+        if (doCluster) {
+          updateCenter()
+          LOG.info(s"task ${taskContext.taskId.index}, center ${center.mkString(",")}")
+          points.clear()
+          previousNumber += currentNumber
+          currentNumber = 0
+        }
+        val newPoint = dataQueue.take()
```

Comment:
this is a blocking call. We suggest against block in Task. You may use "poll" instead.